File: 42.接雨水.go
/*
 * @lc app=leetcode.cn id=42 lang=golang
 *
 * [42] 接雨水
 */
package main

// @lc code=start
func trap(height []int) int {
	res := 0
	n := len(height)
	if n == 0 {
		return res
	}
	leftMax := make([]int, n)
	leftMax[0] = height[0]
	for i := 1; i < n; i++ {
		leftMax[i] = trapMax(leftMax[i-1], height[i])
	}
	rightMax := make([]int, n)
	rightMax[n-1] = height[n-1]
	for i := n - 2; i >= 0; i-- {
		rightMax[i] = trapMax(rightMax[i+1], height[i])
	}
	for i, h := range height {
		res += trapMin(leftMax[i], rightMax[i]) - h
	}
	return res
}

func trapMax(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func trapMin(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @lc code=end

File: 42.接雨水.rs
/*
 * @lc app=leetcode.cn id=42 lang=rust
 *
 * [42] 接雨水
 */

// @lc code=start
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut res = 0;
        let len = height.len();
        if len == 0 {
            return res;
        }
    
        let mut left_max = vec![0; height.len()];
        left_max[0] = height[0];
        for i in 1..len {
            left_max[i] = Ord::max(left_max[i - 1], height[i]);
        }
    
        let mut right_max = vec![0; height.len()];
        right_max[len - 1] = height[len - 1];
        let mut temp_right = len as i32 - 2;
        while temp_right >= 0 {
            right_max[temp_right as usize] = Ord::max(
                right_max[temp_right as usize + 1],
                height[temp_right as usize],
            );
            temp_right -= 1;
        }
        for i in 0..len {
            res += Ord::min(left_max[i], right_max[i]) - height[i];
        }
        res
    }
}
// @lc code=end

